### PR TITLE
NewMainKubelet validates SyncFrequency, minimum GC age, etc.

### PR DESCRIPTION
**Pull request by commonlisp**
NewMainKubelet validates config parameters and propagates error when there is an invalid config parameter. 
Fixes issue #3202.

### CODE REVIEW

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -73,7 +73,16 @@ func NewMainKubelet(
 	maxContainerCount int,
 	sourcesReady SourcesReadyFn,
 	clusterDomain string,
-	clusterDNS net.IP) *Kubelet {
+	clusterDNS net.IP) (*Kubelet, error) {
+	if ri <= 0 {
```
Comment (thockin) [inline pkg/kubelet/kubelet.go]: Can I get you to rename these 2-letter arguments while you're here?

**Comment by thockin** [comment]
Tests failed - they may have run afoul of these new restrictions

**Comment by commonlisp** [comment]
Yes, it does seem to pass 0 for min GC age and max container counts. What would be good defaults here?

**Comment by thockin** [comment]
For tests, I guess it depends on what the test is testing, but something on
the order of the "real" valiues?

**Comment by commonlisp** [comment]
Found the stray MinimumGCAge initialization (defaults to 0) in cmd/kubelet. I did find another inconsistency in the existing codebase: cmd/kubelet defaults syncFrequency to 10s but pkg/standalone#SimpleRunKubelet defaults to 3s. Is this intentional?

**Comment by thockin** [comment]
As I understand SimpleRunKubelet is for tests, so having  faster sync makes
sense.  That said, it was chosen from this air

**Comment by commonlisp** [comment]
Ok, the tests past now and the minor cleanup of variable names is in the commit too (unless we want to separate that for keeping a finer grain history).

File: cmd/kubelet/kubelet.go [hunk 2]
```diff
@@ -57,7 +57,7 @@ var (
 	registryBurst           = flag.Int("registry_burst", 10, "Maximum size of a bursty pulls, temporarily allows pulls to burst to this number, while still not exceeding registry_qps.  Only used if --registry_qps > 0")
 	runonce                 = flag.Bool("runonce", false, "If true, exit after spawning pods from local manifests or remote urls. Exclusive with --etcd_servers and --enable-server")
 	enableDebuggingHandlers = flag.Bool("enable_debugging_handlers", true, "Enables server endpoints for log collection and local running of containers and commands")
-	minimumGCAge            = flag.Duration("minimum_container_ttl_duration", 0, "Minimum age for a finished container before it is garbage collected.  Examples: '300ms', '10s' or '2h45m'")
+	minimumGCAge            = flag.Duration("minimum_container_ttl_duration", 10*time.Second, "Minimum age for a finished container before it is garbage collected.  Examples: '300ms', '10s' or '2h45m'")
```
Comment (thockin) [inline cmd/kubelet/kubelet.go]: too short for a default, IMO.  Should be more like 15 minutes or so

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -61,29 +61,38 @@ type volumeMap map[string]volume.Interface
...
 	minimumGCAge time.Duration,
 	maxContainerCount int,
 	sourcesReady SourcesReadyFn,
 	clusterDomain string,
-	clusterDNS net.IP) *Kubelet {
+	clusterDNS net.IP) (*Kubelet, error) {
+	if resyncInterval <= 0 {
+		return nil, fmt.Errorf("invalid sync frequency %d", resyncInterval)
+	}
+	if minimumGCAge <= 0 {
```
Comment (thockin) [inline pkg/kubelet/kubelet.go]: This is allowed to be 0 in purgeOldest() - you probably should remove that case

File: cmd/kubelet/kubelet.go (hunk 2 above)
Comment (thockin) [inline cmd/kubelet/kubelet.go]: Hmm, given that the default was 0 which mean "instantly" I guess I could live with 1 minute here

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -61,29 +61,38 @@ type volumeMap map[string]volume.Interface
...
+		return nil, fmt.Errorf("invalid minimum GC age %d", minimumGCAge)
+	}
+	if maxContainerCount <= 0 {
```
Comment (thockin) [inline pkg/kubelet/kubelet.go]: I think 0 is valid for this parameter

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (thockin) [inline pkg/kubelet/kubelet.go]: To be clear - either decide that 0 is valid and allow it here, or remove the check for == 0

Comment (thockin) [inline pkg/kubelet/kubelet.go]: you could still check that maxContainers is not < 0, but that's pretty minor

**Comment by thockin** [comment]
LGTM, will merge in AM

**Pull request merged**